### PR TITLE
Tag images with `garden.local.gardener.cloud` instead of `localhost` for local setup

### PR DESCRIPTION
**Pull request by rrhubenov**
**How to categorize this PR?**

/area dev-productivity
/kind enhancement

**What this PR does / why we need it**:
Currently, all images that are built for the local setup are tagged with localhost:5001 as the registry.

This, however, makes it impossible for pods in the cluster to reach the registry since localhost:5001 routes differently for the pods. However, garden.local.gardener.cloud has been setup to route to the cluster node for every pod & node thus allowing pods to reach the registry at port 5001.

There are currently two use cases for this change:
- The Lakom controller needs to communicate with the local registry to pull in digests for every pod. This change will allow us to avoid code that specifically handles this edge case (E.g. when an image is tagged with localhost:5001 treat it as garden.local.gardener.cloud:5001)
- There is already code in gardener that handles this edge case. It can now be removed: https://github.com/gardener/gardener/blob/c5c2e3c613775785ac4056b7ee4e356f53a9f4b1/pkg/utils/oci/helmregistry.go#L92-L96

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```breaking developer
In the local development setup, the images are pushed to `garden.local.gardener.cloud:5001` instead of `localhost:5001` now. Please add `127.0.0.1 garden.local.gardener.cloud` to your `/etc/hosts`. 
```

### CODE REVIEW

**Comment by vpnachev** [comment]
Most probably the following line will need to be adjusted 
https://github.com/gardener/gardener/blob/78517a560d42db68f6b1161b473d9fb6b676c968/docs/deployment/getting_started_locally.md?plain=1#L69

I.e. `garden.local.gardener.cloud` should resolve to both `127.0.0.1` and `::1`.

File: pkg/utils/oci/helmregistry.go [hunk 1]
```diff
@@ -89,11 +89,10 @@ func buildRef(oci *gardencorev1.OCIRepository) (name.Reference, error) {
...
 	}
 
-	// in the local setup, we need to replace the registry and configure that we don't want to use TLS
-	if strings.Contains(ref, localRegistry) {
-		ref = strings.Replace(ref, localRegistry, inKubernetesRegistry, 1)
+	// in the local setup we don't want to use TLS
+        if strings.Contains(ref, "garden.local.gardener.cloud:5001") {
 		opts = append(opts, name.Insecure)
-	}
+        }
```
Comment (vpnachev) [inline pkg/utils/oci/helmregistry.go]: Use tabs instead of spaces here, please.

File: pkg/utils/oci/helmregistry.go [hunk 2]
```diff
@@ -89,11 +89,10 @@ func buildRef(oci *gardencorev1.OCIRepository) (name.Reference, error) {
 		name.StrictValidation,
 	}
 
-	// in the local setup, we need to replace the registry and configure that we don't want to use TLS
-	if strings.Contains(ref, localRegistry) {
-		ref = strings.Replace(ref, localRegistry, inKubernetesRegistry, 1)
+	// in the local setup we don't want to use TLS
+        if strings.Contains(ref, "garden.local.gardener.cloud:5001") {
```
Comment (vpnachev) [inline pkg/utils/oci/helmregistry.go]: You can use the `inKubernetesRegistry` const here
```suggestion
	if strings.Contains(ref, inKubernetesRegistry) {
```

**Review by vpnachev — COMMENTED** [review]
Thanks, overall looks good!

I have just few very minor nits and doc suggestions.
Please, mention that the registry can be accessed also via `garden.local.gardener.cloud` on the following places:
- https://github.com/gardener/gardener/blob/2f5ad80b0090983d4873673c1dc92ad7102be099/docs/deployment/getting_started_locally_with_extensions.md?plain=1#L88
- https://github.com/gardener/gardener/blob/2f5ad80b0090983d4873673c1dc92ad7102be099/docs/deployment/getting_started_locally.md?plain=1#L53

File: hack/kind-up.sh [hunk 3]
```diff
@@ -58,7 +58,24 @@ parse_flags() {
...
+  elif [[ "$OSTYPE" == "linux-gnu"* ]]; then
+    glgc_ip_address=$(getent hosts garden.local.gardener.cloud | cut -d' ' -f1)
+  else
+    echo "WARN: Unknown OS. Make sure garden.local.gardener.cloud resolves to 127.0.0.1"
+  fi
+
+  if [ "$glgc_ip_address" != "127.0.0.1" ]; then
+      echo "garden.local.gardener.cloud does not resolve to 127.0.0.1. Please add a line for it in /etc/hosts"
+      exit 1
+  fi
```
Comment (vpnachev) [inline hack/kind-up.sh]: The check seems to not belong to the `setup_kind_network` function, therefore I would propose to move it to dedicated function, e.g. `check_local_dns_records` or something similar and just call it before or after `check_shell_dependencies`.

File: hack/kind-up.sh [hunk 4]
```diff
@@ -58,7 +58,24 @@ parse_flags() {
...
+        glgc_ip_address=$(dscacheutil -q host -a name garden.local.gardener.cloud | grep "ip_address" | head -n 1| cut -d' ' -f2)
+    fi
+  elif [[ "$OSTYPE" == "linux-gnu"* ]]; then
+    glgc_ip_address=$(getent hosts garden.local.gardener.cloud | cut -d' ' -f1)
+  else
+    echo "WARN: Unknown OS. Make sure garden.local.gardener.cloud resolves to 127.0.0.1"
+  fi
+
+  if [ "$glgc_ip_address" != "127.0.0.1" ]; then
+      echo "garden.local.gardener.cloud does not resolve to 127.0.0.1. Please add a line for it in /etc/hosts"
```
Comment (vpnachev) [inline hack/kind-up.sh]:
```suggestion
      echo "Error: garden.local.gardener.cloud does not resolve to 127.0.0.1. Please add a line for it in /etc/hosts"
```

to make it similar to how other errors are handled in the script.

File: hack/kind-up.sh [hunk 5]
```diff
@@ -58,7 +58,24 @@ parse_flags() {
...
+  glgc_ip_address=""
+    
+  if [[ "$OSTYPE" == "darwin"* ]]; then
+    if dscacheutil -q host -a name garden.local.gardener.cloud | grep -q "ip_address"; then
+        glgc_ip_address=$(dscacheutil -q host -a name garden.local.gardener.cloud | grep "ip_address" | head -n 1| cut -d' ' -f2)
+    fi
+  elif [[ "$OSTYPE" == "linux-gnu"* ]]; then
+    glgc_ip_address=$(getent hosts garden.local.gardener.cloud | cut -d' ' -f1)
+  else
+    echo "WARN: Unknown OS. Make sure garden.local.gardener.cloud resolves to 127.0.0.1"
```
Comment (vpnachev) [inline hack/kind-up.sh]:
```suggestion
    echo "Warning: Unknown OS, make sure garden.local.gardener.cloud resolves to 127.0.0.1"
```

File: pkg/utils/oci/helmregistry.go [hunk 6]
```diff
@@ -89,9 +89,8 @@ func buildRef(oci *gardencorev1.OCIRepository) (name.Reference, error) {
 		name.StrictValidation,
 	}
 
-	// in the local setup, we need to replace the registry and configure that we don't want to use TLS
-	if strings.Contains(ref, localRegistry) {
-		ref = strings.Replace(ref, localRegistry, inKubernetesRegistry, 1)
+	// in the local setup we don't want to use TLS
+	if strings.Contains(ref, inKubernetesRegistry) {
```
Comment (vpnachev) [inline pkg/utils/oci/helmregistry.go:93]: You have to adapt the unit test https://github.com/gardener/gardener/blob/78565916bef348d472d635e0e1b2449ed6b335d9/pkg/utils/oci/helmregistry_test.go#L126-L149

**Review by vpnachev — COMMENTED** [review]
It should be checked why the other tests are failing, I see nothing obvious from the available output.

**Comment by rrhubenov** [comment]
> It should be checked why the other tests are failing, I see nothing obvious from the available output.

The issue is twofold -- my code fails when getent returns a non-zero exit code which I'll fix in the next commit, but the bigger issue is that I need to modify the cicd setup code to add an entry for garden.local.gardener.cloud in the /etc/hosts file for the host machine running the e2e tests.

File: hack/kind-up.sh [hunk 7]
```diff
@@ -53,6 +53,27 @@ parse_flags() {
...
+  elif [[ "$OSTYPE" == "linux-gnu"* ]]; then
+    # Suppress exit code using "|| true"
+    glgc_ip_address="$(getent hosts garden.local.gardener.cloud | cut -d' ' -f1 || true)"
+  else
+    echo "Warning: Unknown OS. Make sure garden.local.gardener.cloud resolves to 127.0.0.1"
+    return 0
+  fi
+
+  if [ "$glgc_ip_address" != "127.0.0.1" ]; then
+      echo "Error: garden.local.gardener.cloud does not resolve to 127.0.0.1. Please add a line for it in /etc/hosts"
```
Comment (oliver-goetz) [inline hack/kind-up.sh:72]: If the e2e tests should run in prow, you have to add this resolution to `/etc/hosts`.
We are doing something similar for the DNS names of the shoots [here](https://github.com/gardener/gardener/blob/00b41a217ff62f25842e4ed53a0b99b319ecec84/hack/test-e2e-local.sh#L105-L119). The existence of `CI` and `ARTIFACTS` envs can tell you that the test is running in prow.

**Review by oliver-goetz — COMMENTED** [review]
Thanks for your PR 👍 
I have a suggestion how to fix the e2e tests.

File: hack/kind-up.sh (hunk 7 above)
Comment (rrhubenov) [reply hack/kind-up.sh:72]: Thank you! I haven't had the time to look into it, so you just saved me some time. I'll finish this up soon

File: hack/push-helm.sh [hunk 8]
```diff
@@ -31,4 +31,4 @@ yq -i "$image_ref |= \"$IMG\"" "$chart_dir/values.yaml"
 yq -i ".name |= \"$name\"" "$chart_dir/Chart.yaml"
 
 helm package "$chart_dir" -d "$chart_dir" --version "$tag"
-helm push "$chart_dir/$name-$tag.tgz" "oci://$registry"
+helm push --plain-http "$chart_dir/$name-$tag.tgz" "oci://$registry"
```
Comment (oliver-goetz) [inline hack/push-helm.sh:34]: This doesn't seem to work as expected. See the errors in the tests ([ref](https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/10257/pull-gardener-e2e-kind-operator/1835927403969384448#1:build-log.txt%3A210-211), [ref](https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/10257/pull-gardener-e2e-kind-operator-seed/1835927403130523648#1:build-log.txt%3A211))

Comment (rrhubenov) [reply hack/push-helm.sh:34]: Hi Oliver, it turns out I had forgotten to add `garden.local.gardener.cloud:5001` as an insecure registry to the other two skaffold pipeline files. There's one last issue I'm trying to figure out and then I'll clean up this PR for merging

File: docs/deployment/getting_started_locally.md [hunk 9]
```diff
@@ -37,8 +37,6 @@ In these cases, you might want to check out one of the following options that ru
 
 ```bash
 make kind-up
-```
-
```
Comment (vpnachev) [inline docs/deployment/getting_started_locally.md]: Kindly revert this change.

File: hack/kind-up.sh [hunk 10]
```diff
@@ -53,6 +53,27 @@ parse_flags() {
...
+    glgc_ip_address=$(dscacheutil -q host -a name garden.local.gardener.cloud | grep "ip_address" | head -n 1| cut -d' ' -f2 || true)
+  elif [[ "$OSTYPE" == "linux-gnu"* ]]; then
+    # Suppress exit code using "|| true"
+    glgc_ip_address="$(getent ahosts garden.local.gardener.cloud | cut -d' ' -f1 || true)"
+  else
+    echo "Warning: Unknown OS. Make sure garden.local.gardener.cloud resolves to 127.0.0.1"
+    return 0
+  fi
+    
+  if ! echo "$glgc_ip_address" | grep "127.0.0.1" ; then
```
Comment (vpnachev) [inline hack/kind-up.sh]: You can suppress the output here, otherwise the matched results will be printed on the stdout.

```suggestion
  if ! echo "$glgc_ip_address" | grep -q "127.0.0.1" ; then
```

**Review by vpnachev — COMMENTED** [review]
Just one minor nit and one suggestion, rest is looking good to me.

**Comment by rrhubenov** [comment]
Thank you @vpnachev, very good catches

File: hack/ci-e2e-kind-ha-multi-zone.sh [hunk 11]
```diff
@@ -11,6 +11,12 @@ set -x
 
 source $(dirname "${0}")/ci-common.sh
 
+# If running in prow, we need to ensure that garden.local.gardener.cloud resolves to localhost
+if [ -n "${CI:-}" -a -n "${ARTIFACTS:-}" ]; then
+    printf "\n127.0.0.1 garden.local.gardener.cloud\n" >> /etc/hosts
+    printf "\n::1 garden.local.gardener.cloud\n" >> /etc/hosts
+fi
```
Comment (oliver-goetz) [inline hack/ci-e2e-kind-ha-multi-zone.sh]: These lines have been added to all `ci-e2e-*.sh` files. Instead of copying it everywhere, please create a function in `hack/ci-common.sh` and call it here. The file is sourced in all the scripts anyway.

**Review by oliver-goetz — COMMENTED** [review]
Thank you 🙂 
I just found one last thing.

**Review by oliver-goetz — APPROVED** [review]
Awesome, thanks 😄 

/lgtm
/approve
